torch/fx/experimental/symbolic_shapes.py

```diff
@@ -274,6 +279,20 @@ def parallel_and(*args):
         return False
     return all(args)
 
+def sym_eq(x, y):
+    """
+    Like ==, but when run on list/tuple, it will recursively test equality
+    and use sym_and to join the results together, without guarding.
+    """
+    if (isinstance(x, tuple) and isinstance(y, tuple)) or (isinstance(x, list) and isinstance(y, list)):
+        if len(x) != len(y):
+            return False
+        return functools.reduce(operator.and_, map(sym_eq, x, y), True)
```

Comment:
I kind of thought that the default implementation of `==` would do this for lists, but I guess that's wrong

Reply:
The default == implementation is short circuiting: 

```
>>> class A:
...   def __eq__(self, other):
...     assert False
... 
>>> 
>>> [1, A()] == [2, A()]
False
>>> [1, A()] == [1, A()]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in __eq__
AssertionError
```